### PR TITLE
mgr/dashboard: add runtime type checks

### PR DESCRIPTION
**Pull request by epuertat**
Adds a new dependency ([typeguard](https://github.com/agronholm/typeguard/)) to perform runtime typechecks by means of the '@typechecker'.
    
[Available packages](https://pkgs.org/search/?q=typeguard) for:
- Ubuntu >=20
- Fedora >=32
- CentOS 8 via EPEL 8
- OpenSUSE Tumbleweed

```bash
# changing input arg from "ids" to "id"
# curl -X PUT "https://localhost:11000/api/osd/flags/individual" ... -d "{\"flags\":{\"noout\":true,\"noin\":true,\"noup\":true,\"nodown\":true},\"id\":[0]}"
{
  "status": "400 Bad Request",
  "detail": "missing a required argument: 'ids'",
  "request_id": "5890bc08-44a3-4627-845a-0893a7d037d2"
}
```

```bash
# adding a wrong extra field to "flags": "noblah"
# curl -X PUT "https://localhost:11000/api/osd/flags/individual" ... -d "{\"flags\":{\"noout\":true,\"noin\":true,\"noup\":true,\"nodown\":true,\"noblah\":false},\"ids\":[0]}"
{
  "status": "400 Bad Request",
  "detail": "extra key(s) (\"noblah\") in argument \"flags\"",
  "request_id": "5c2575ec-2a38-4c84-9d92-f19174f36853"
}
```

Fixes: https://tracker.ceph.com/issues/46423
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
downside it that this introduces yet another dependency we need to fulfill.

File: src/pybind/mgr/dashboard/controllers/osd.py [hunk 1]
```diff
@@ -469,7 +477,8 @@ def bulk_set(self, flags):
                                     'to.')
                  },
                  responses={200: EXPORT_INDIV_FLAGS_SCHEMA})
-    def set_individual(self, flags, ids):
+    @typechecked
+    def set_individual(self, flags: IndividualFlags, ids: List[int]):
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/controllers/osd.py:481]: if this API is only consumed by ourselves, i.e. it's an internal API. do we still need to install `typeguard` in production environment, or we just need to install it in our testing environment, when running teuthology based test and `make check` tests?

File: src/pybind/mgr/dashboard/controllers/osd.py [hunk 2]
```diff
@@ -46,6 +48,11 @@
     "flags": ([str], "List of active flags")
 }
 
+class IndividualFlags(TypedDict):
+    noout: bool
+    noin: bool
+    noup: bool
+    nodown: bool
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/controllers/osd.py:55]: might want to move this change into another commit.

File: src/pybind/mgr/dashboard/plugins/debug.py [hunk 3]
```diff
@@ -80,10 +81,11 @@ def handler(self, action: Actions):
 
     def custom_error_response(self, status, message, traceback, version):
         self.response.headers['Content-Type'] = 'application/json'
-        error_response = dict(status=status, detail=message, request_id=str(self.request.unique_id))
+        error_response = dict(status=status, detail=html.unescape(message),
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/plugins/debug.py:84]: this change deserves a separate commit.

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/controllers/osd.py (hunk 1 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/osd.py:482]: @tchaikov The idea of this improvement is to be on production code. Duck-typing is great most of the time, but if you're expecting a list of labels for a host, but the API user sends just a single label, both input args (`list` and `str`) are iterable, but in one case you get a host labelled liked `['rgw', 'mds']`, while in the other `['r', 'g', 'w']`... So we need to actively enforce typecheck in methods. We could populate with `asserts` and maintain the assertions in sync with the type hints... but that's what `typeguard` does for us 🥳

File: src/pybind/mgr/dashboard/plugins/debug.py (hunk 3 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/plugins/debug.py:84]: You're rigth! This was mostly exploratory work.

**Comment by epuertat** [comment]
jenkins retest this

**Comment by epuertat** [comment]
jenkins retest this please

**Comment by epuertat** [comment]
jenkins restest this please

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
actually I'm :-1: for using typeguard, as it doesn't provide a way to generate an OpenAPIV3Spec.  If we're introducing this, we should aim for solving more than this problem. +1 for using pydantic.

**Comment by epuertat** [comment]
> actually I'm  for using typeguard, as it doesn't provide a way to generate an OpenAPIV3Spec. If we're introducing this, we should aim for solving more than this problem. +1 for using pydantic.

@sebastian-philipp when do you plan to submit that PR with pydantic?

**Comment by epuertat** [comment]
> @epuertat we were looking at this in the RADOS PR scrub; is it still needed?

Some improved type safety in Python will always be welcome... but we can close this one.
